Storage: Add Oracle Cloud Infrastructure Object Storage Bucket support (#4661)
thanos-io/thanos: docs/storage.md

  directory: ""
```

### OCI

[Oracle Cloud Infrastructure](https://www.oracle.com/cloud). To configure OCI Object Storage as Thanos Object Store, you need access to OCI Object Storage.
You can get a free OCI Cloud Services by following the [Oracle Cloud Free Tier](https://www.oracle.com/cloud/free) page.
More information about [Overview of Object Storage](https://docs.cloud.oracle.com/en-us/iaas/Content/Object/Concepts/objectstorageoverview.htm).

OCI Object Storage supports three different providers, default, instance principle, and raw. 
The default HTTP Client comes with this Thanos OCI Object Storage should be able to handle most of the use cases.

> By default, like other clients, Thanos OCI Object Storage client will storage the data from a single prometheus instance to the top directory of the bucket. 

Author: I will fix it in my next push (I may wait for your feedback from other comments first).

> By default, like other clients, Thanos OCI Object Storage client will storage the data from a single prometheus instance to the top directory of the bucket. 

Author: Pushed the fix.

> This means each bucket can only store one prometheus instance data.

Author: Sorry for the poor description.
When thanos uploads all the blocks from a prometheus instance, it will create a directory structure similar to the following in OCI objectstore
```text
|-- 01FF8QX35297NG9HCDHGH1BN1
|   |-- chunks
|       |-- 000001
|       |-- 000002
|       |-- ....
|-- 01FFPFKMAQQNF4N945X7JWPQ8
|   |-- chunks
|.      |-- ....
|-- .... other prometheus blocks
|
|-- debug
    |-- meta
        |-- .....
```
`I consider above as one prometheus instance data.`

In this PR, a new option called object_base_path is added.
If defined, the whole directory structure above will be uploaded under the object_base_path prefix.

In second screenshot in this PR, it holds 4 prometheus instances.
Prometheus A - object_base_path is set to thanos.dev.thanos
Prometheus B - object_base_path is set to thanos.meta.sa/0
Prometheus C - object_base_path is set to thanos.meta.sa/1
Prometheus D - object_base_path is set to thanos.meta.sa/2

Here is the illustration in OCI objectstorage bucket.
```
|-- thanos.dev.thanos
    |--  <<Promethus-A with all blocks and debug meta>>
|-- thanos.meta.sa
    |-- 0 <<Promethus-B with all blocks and debug meta>>
    |-- 1 <<Promethus-C with all blocks and debug meta>>
    |-- 2 <<Promethus-D with all blocks and debug meta>>
        |-- 01FF8QX35297NG9HCDHGH1BN1
        |   |-- chunks
        |       |-- 000001
        |       |-- 000002
        |       |-- ....
        |-- 01FFPFKMAQQNF4N945X7JWPQ8
        |   |-- chunks
        |.      |-- ....
        |-- .... other prometheus blocks
        |
        |-- debug
            |-- meta
                |-- .....
```

If you want to provide a better description for the Thanos community, I will be happy to change it in my PR.

> This means each bucket can only store one prometheus instance data.

Reviewer: Yes, that's what I mean - this https://github.com/thanos-io/thanos/issues/1318 has been a long-standing issue. I think this kind of functionality needs to be added "one level above" i.e. not at the raw object storage calls level. Otherwise, it would be weird to have duplicate functionality. I think others would agree as well. My suggestion would be to remove this prefix functionality and move it one level above so that it could be used with any kind of object storage. It could be done in a separate pull request. We could double check in #thanos-dev, though

> This means each bucket can only store one prometheus instance data.

Author: Got it. Yes, adding the feature to `one level up` is a great idea for all Objectstorage clients.

Since allowing the same bucket to hold multiple prometheus instances' data is the main use case of our team and we would like to use Thanos master branch instead of my forked branch.

I can think of three different approaches (order by my preferences) to avoid duplicate functionality in my PR:
**_Approach # 1_**
- hide the object_base_path configuration from storage.md documentation
- without the object_base_path setting, my PR will act exactly the same as other ObjecctStorage Clients
- the hidden feature can still be used by the insider who know the existence of this feature
- after the code check it, whoever want to work on #1318 will have an example how OCI client is implemented.

**_Approach # 2_**
- hide the object_base_path configuration from storage.md documentation
- hard code the object_base_path to empty string and won't accept any hidden customization

**_Approach # 3_**
Remove all the object_base_path throughout the PR.

Please let me know which approach is acceptable for you.
Thanks!

> This means each bucket can only store one prometheus instance data.

Reviewer 2: Sorry for the delay again, bit busy times. 
As I like the different approaches I'm not that fan of hidden features in a way that one needs to know the existence in an open source project. Also I took some time to read the long standing issue and the consensus after the discussion was this:

> Looks like the decision is that we are ok for simple custom prefix on config level (for all objstorages) in YAML (:

So, I would say that if we implement it, it should be "that". Still open for discussion though, but then it should require a lot more discussion from everyone else again IMO ^^

I would say, let's just implement this on level higher as discussed in that main issue. For all the storage backends, implement at least the config in a way that it can say "This is not supported yet for this object store" if one would configure it. With then the exception for OCI, in which you did implement it. 

This gives a bit more motivation to get that issue fixed, and we can have iterations per objectstorage to actually implement it. I think by saying "It needs to be done for them all in one go", is really limiting us to get it moving forward. 

What do you think? Also, I'm actually happy to help on that. I got some spare time soon.

> This means each bucket can only store one prometheus instance data.

Reviewer: I really think we should do (3) approach here. Then, you could implement it in a follow up PR. We have interfaces for object storage so IMHO it should be fairly easy to implement prefix functionality.

> This means each bucket can only store one prometheus instance data.

Author: I will remove the prefix option when I have a chance.

> This means each bucket can only store one prometheus instance data.

Reviewer: Don't forget to remove this sentence as it is not true in practice, it is _very_ hard to get a UUID collision.

To support multiple prometheus instances with a single bucket, a new `object_base_path` option is introduced in OCI Object Storage client. 
If defined, the value will be used as the prefix of the prometheus data in OCI bucket.

For example, if you want to use the same bucket for 2 prometheus instances, you can define `object_base_path` as `/path/for/prometheus1` 
for the first prometheus instance, and `/path/for/prometheus2` for the second prometheus instance.

#### Default Provider
The default config provider will look for configurations in 3 places: file in `$HOME/.oci/config`, `$HOME/.obmcs/config`, and

> variables names starting with the string TF_VAR. If the same configuration is found in multiple places the provider will

Author: Yes, TF_VAR is referring to the OCI environment variables that are used to connect to OCI.
For users who use OCI, they should be very familiar with the name TF_VAR. Probably, TF_VAR will be the first thing that they learn when using OCI outside of OCI Console.

This is the verbatim from the public OCI GO SDK APIs. For this PR, I tried to be consistent with the definition in OCI GO SDK so the users could feel like home.

See https://github.com/oracle/oci-go-sdk/blob/master/common/client.go#L301 for details.

> variables names starting with the string TF_VAR. If the same configuration is found in multiple places the provider will

Reviewer: :+1:

> prefer the first one.

Author: See reply from above for full `Default Provider` explanation.

For Example:
```yaml
type: OCI
config:
  provider: "default"
  bucket: ""
  compartment_ocid: ""

>   object_base_path: ""            // Optional object base path (prefix) in OCI bucket.

Author: Please see the explanation above for  _**What does this mean? Why only one?**_
Hope it will make more sense after you read my explanation.

If you want to use different variable name or an alternative approach for this new objectstorage prefix feature, please let me know.

    max_idle_conns: 100           // Optional maximum number of idle (keep-alive) connections across all hosts. Zero means no limit.
    max_idle_conns_per_host: 100  // Optional maximum idle (keep-alive) connections to keep per-host. If zero, DefaultMaxIdleConnsPerHost=2 is used.
    max_conns_per_host: 0         // Optional maximum total number of connections per host.
    disable_compression: false    // Optional. If true, prevents the Transport from requesting compression.
```

#### Instance Principle Provider
For Example:
```yaml
type: OCI
config:
  provider: "instance-principal"
  bucket: ""
  compartment_ocid: ""
```
You can also include any of the optional configuration just like the example in `Default Provider`. 

#### Raw Provider
For Example:
```yaml
type: OCI
config:
  provider: "raw"
  bucket: ""
  compartment_ocid: ""
  tenancy_ocid: ""
  user_ocid: ""
  region: ""
  fingerprint: ""
  privatekey: ""
  passphrase: ""         // Optional passphrase to encrypt the private API Signing key
```
You can also include any of the optional configuration just like the example in `Default Provider`.

> #### TestObjStore_AcceptanceTest_e2e

Author: I will remove this section in my next push (I will wait for your feedback from other comments first).

> #### TestObjStore_AcceptanceTest_e2e

Author: Pushed the fix.

  directory: ""
```

### OCI

[Oracle Cloud Infrastructure](https://www.oracle.com/cloud). To configure OCI Object Storage as Thanos Object Store, you need access to OCI Object Storage.
You can get a free OCI Cloud Services by following the [Oracle Cloud Free Tier](https://www.oracle.com/cloud/free) page.
More information about [Overview of Object Storage](https://docs.cloud.oracle.com/en-us/iaas/Content/Object/Concepts/objectstorageoverview.htm).

> OCI Object Storage supports three different providers, default, instance principle, and raw. 

Reviewer 3: This should be instance _principal_.

Author: Thanks for the eagle eyes. Push the fix a minute ago.

    response_header_timeout: 2m   // Optional amount of time to wait for a server's response headers after fully writing the request.
    tls_handshake_timeout: 10s    // Optional maximum amount of time waiting to wait for a TLS handshake. Zero means no timeout.
    expect_continue_timeout: 1s   // Optional amount of time to wait for a server's first response headers. Zero means no timeout and causes the body to be sent immediately.
    insecure_skip_verify: false   // Optional. If true, crypto/tls accepts any certificate presented by the server and any host name in that certificate.
    max_idle_conns: 100           // Optional maximum number of idle (keep-alive) connections across all hosts. Zero means no limit.
    max_idle_conns_per_host: 100  // Optional maximum idle (keep-alive) connections to keep per-host. If zero, DefaultMaxIdleConnsPerHost=2 is used.
    max_conns_per_host: 0         // Optional maximum total number of connections per host.
    disable_compression: false    // Optional. If true, prevents the Transport from requesting compression.
```

> #### Instance Principle Provider

Reviewer 3: This should be _Principal_ as per the `provider` value below. :)

Author: Same, pushed the fix a minute ago.

  directory: ""
```

### OCI - Oracle Cloud Infrastructure

To configure OCI Object Storage as Thanos Object Store, you need access to OCI Object Storage which supports three different providers, default, instance principal, and raw. 

> The default HTTP Client comes with this Thanos OCI Object Storage should be able to handle most of the use cases.

Reviewer 2: Let's try to make this more like a statement. Does it work or not? As a user, you don't want the "should" and "most of the use-cases" :D 

If it does work, we can rename this part towards something like "OCI has 3 different providers; x, y, and z, which are supported by this object storage client implementation" - Or something like that. 

It could also be I'm misreading this part, so feel free to correct me on that.

If there are any docs, I still think it's good to link towards that as well.

Author: Make sense Just pushed an updated.

  directory: ""
```

### OCI

[Oracle Cloud Infrastructure](https://www.oracle.com/cloud). To configure OCI Object Storage as Thanos Object Store, you need access to OCI Object Storage.

> You can get a free OCI Cloud Services by following the [Oracle Cloud Free Tier](https://www.oracle.com/cloud/free) page.

Reviewer 2: IMO, it's fine to link to docs like on line 477, 479, but this is more promotion / advertising. Let's keep this project neutral in that sense (i.e. we also don't add things about other cloud vendors free tiers).

Author: Hi @wiardvanrij 
Understand. I just pushed another commit with updated description. Please take a look. Thanks for reviewing it.

Reviewer 2: Thanks, added a new review. I will also test it soon btw, just haven't got the time for it yet.

Reviewer 3: @wiardvanrij we have an OCI tenancy available for CNCF project testing purposes, so if you can point me towards your test framework, we can start looking into providing the resources for CI-triggered automatic testing on an ongoing basis too.

prefix: ""
```

> ### OCI - Oracle Cloud Infrastructure

Author: Pushed the updates.

To configure OCI Object Storage as Thanos Object Store, you need access to OCI Object Storage.

> OCI has 3 different providers; default, instance principal, and raw, which are supported by this object storage client implementation

Author: Pushed the updates.

> #### Default Provider
> The default config provider will look for configurations in 3 places: file in `$HOME/.oci/config`, `$HOME/.obmcs/config`, and
> variables names starting with the string TF_VAR. If the same configuration is found in multiple places the provider will
> prefer the first one.

Reviewer 3: ```suggestion
#### API Signing Key

The default API signing key authentication provider leverages same [configuration as the OCI CLI](https://docs.oracle.com/en-us/iaas/Content/API/Concepts/cliconcepts.htm) which is usually stored in at `$HOME/.oci/config` or via variable names starting with the string `OCI_CLI`. You can also use environment variables that start with `TF_VAR`. If the same configuration is found in multiple places the provider will prefer the first one.
```

Author: Pushed the updates.

> For Example:

Reviewer 3: ```suggestion
The following example configures the provider to look for an existing API signing key for authentication:

```

Author: Pushed the updates.

    response_header_timeout: 2m   // Optional amount of time to wait for a server's response headers after fully writing the request.
    tls_handshake_timeout: 10s    // Optional maximum amount of time waiting to wait for a TLS handshake. Zero means no timeout.
    expect_continue_timeout: 1s   // Optional amount of time to wait for a server's first response headers. Zero means no timeout and causes the body to be sent immediately.
    insecure_skip_verify: false   // Optional. If true, crypto/tls accepts any certificate presented by the server and any host name in that certificate.
    max_idle_conns: 100           // Optional maximum number of idle (keep-alive) connections across all hosts. Zero means no limit.
    max_idle_conns_per_host: 100  // Optional maximum idle (keep-alive) connections to keep per-host. If zero, DefaultMaxIdleConnsPerHost=2 is used.
    max_conns_per_host: 0         // Optional maximum total number of connections per host.
    disable_compression: false    // Optional. If true, prevents the Transport from requesting compression.
    client_timeout: 90s           // Optional time limit for requests made by the HTTP Client.
```

> 
> 

Reviewer 3: ```suggestion

Alternatively, you can provide API signing key authentication parameters in the configuration:

```yaml
type: OCI
config:
  provider: "raw"
  bucket: "thanos" // name of the object storage bucket
  compartment_ocid: "ocid1.compartment.oc1..."  // compartment in which the object storage bucket was created
  tenancy_ocid: "ocid1.tenancy.oc1..."
  user_ocid: "ocid1.user.oc1..."
  region: "us-ashburn-1"
  fingerprint: "aa:a1:bb:b1:cc:c1:dd:d1:ee:e1:ff:f1:aa:a2:bb:b2"
  privatekey: "/path/to/oci_api_key.pem"
  passphrase: ""         // Optional passphrase used to encrypt the private API Signing key

```

> #### Instance Principal Provider
> For Example:

Reviewer 3: ```suggestion
#### Instance Principal authentication

Instance principal authentication is available when the provider is running on a compute instance within Oracle Cloud Infrastructure.  Each compute instance has its own identity, and it authenticates using the certificates that are added to it. These certificates are automatically created, assigned to instances and rotated, preventing the need for you to distribute credentials to your hosts and rotate them. You are required to [configure an appropriate Dynamic Group and assign that group permissions via policies](https://docs.oracle.com/en-us/iaas/Content/Identity/Tasks/callingservicesfrominstances.htm) so that the compute instance is authorised to access the object storage bucket.

The following configuration example enables instance principal authentication:
```

```yaml
type: OCI
config:
  provider: "instance-principal"
  bucket: ""
  compartment_ocid: ""
```
You can also include any of the optional configuration just like the example in `Default Provider`. 

#### Raw Provider
For Example:
```yaml
type: OCI
config:
  provider: "raw"
  bucket: ""
  compartment_ocid: ""
  tenancy_ocid: ""
  user_ocid: ""
  region: ""
  fingerprint: ""
  privatekey: ""
  passphrase: ""         // Optional passphrase to encrypt the private API Signing key
```

> You can also include any of the optional configuration just like the example in `Default Provider`.

Reviewer 3: This block should be removed.
